File: src/main/java/net/nodus/core/site/SiteKeyService.java
package net.nodus.core.site;

import java.security.SecureRandom;
import java.time.Instant;
import java.util.Base64;
import java.util.Objects;
import net.nodus.global.common.exception.GlobalException;
import net.nodus.core.site.dto.IssuedSiteKey;
import net.nodus.database.site.SiteKeyRepository;
import net.nodus.database.site.SiteKeyStatus;
import net.nodus.database.site.SiteRepository;
import net.nodus.database.site.entity.Site;
import net.nodus.database.site.entity.SiteKey;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class SiteKeyService {

    private static final int KEY_PREFIX_LENGTH = 24;

    private final SiteKeyRepository siteKeyRepository;
    private final SiteRepository siteRepository;
    private final BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();
    private final SecureRandom secureRandom = new SecureRandom();

    public SiteKeyService(SiteKeyRepository siteKeyRepository, SiteRepository siteRepository) {
        this.siteKeyRepository = siteKeyRepository;
        this.siteRepository = siteRepository;
    }

    public IssuedSiteKey issue(Site site) {
        String siteId = Objects.requireNonNull(site.getId());

        revokeActive(siteId);

        String rawKey = generateRawKey();
        SiteKey siteKey = siteKeyRepository.save(new SiteKey(
            site.getUserAccountId(),
            siteId,
            rawKey.substring(0, KEY_PREFIX_LENGTH),
            Objects.requireNonNull(passwordEncoder.encode(rawKey))
        ));

        return new IssuedSiteKey(siteKey, rawKey);
    }

    public SiteKey authenticate(String rawKey) {
        var candidates = siteKeyRepository.findAllByKeyPrefixAndStatusAndDeletedAtIsNull(
            rawKey.substring(0, Math.min(KEY_PREFIX_LENGTH, rawKey.length())),
            SiteKeyStatus.ACTIVE
        );

        SiteKey siteKey = candidates.stream()
            .filter(candidate -> passwordEncoder.matches(rawKey, candidate.getKeyHash()))
            .findFirst()
            .orElseThrow(() -> new GlobalException.Unauthorized("Invalid key"));

        siteRepository
            .findByIdAndUserAccountIdAndDeletedAtIsNull(siteKey.getSiteId(),
                siteKey.getUserAccountId())
            .orElseThrow(() -> new GlobalException.DataNotFound("Site not found"));

        return siteKey;
    }

    public void revokeActive(String siteId) {
        siteKeyRepository.findBySiteIdAndStatusAndDeletedAtIsNull(siteId, SiteKeyStatus.ACTIVE)
            .ifPresent(activeKey -> {
                activeKey.setStatus(SiteKeyStatus.REVOKED);
                activeKey.setRevokedAt(Instant.now());
                siteKeyRepository.save(activeKey);
            });
    }

    private String generateRawKey() {
        byte[] bytes = new byte[32];
        secureRandom.nextBytes(bytes);

        return "nodus_live_" + Base64.getUrlEncoder()
            .withoutPadding()
            .encodeToString(bytes);
    }
}

File: src/main/kotlin/net/nodus/site/SiteKeyService.kt
package net.nodus.site

import net.nodus.common.exception.GlobalException
import net.nodus.site.dto.IssuedSiteKey
import net.nodus.site.entity.Site
import net.nodus.site.entity.SiteKey
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service
import java.security.SecureRandom
import java.time.Instant
import java.util.Base64

@Service
class SiteKeyService(
    private val siteKeyRepository: SiteKeyRepository,
    private val siteRepository: SiteRepository,
) {
    private val passwordEncoder = BCryptPasswordEncoder()

    fun issue(site: Site): IssuedSiteKey {
        val siteId = requireNotNull(site.id)

        revokeActive(siteId)

        val rawKey = generateRawKey()
        val siteKey = siteKeyRepository.save(
            SiteKey(
                userAccountId = site.userAccountId,
                siteId = siteId,
                keyPrefix = rawKey.take(KEY_PREFIX_LENGTH),
                keyHash = requireNotNull(passwordEncoder.encode(rawKey)),
            )
        )

        return IssuedSiteKey(
            siteKey = siteKey,
            rawKey = rawKey,
        )
    }

    fun authenticate(rawKey: String): SiteKey {
        val candidates = siteKeyRepository.findAllByKeyPrefixAndStatusAndDeletedAtIsNull(
            keyPrefix = rawKey.take(KEY_PREFIX_LENGTH),
            status = SiteKeyStatus.ACTIVE,
        )

        val siteKey = candidates.firstOrNull {
            passwordEncoder.matches(rawKey, it.keyHash)
        } ?: throw GlobalException.Unauthorized("Invalid key")

        siteRepository.findByIdAndUserAccountIdAndDeletedAtIsNull(
            id = siteKey.siteId,
            userAccountId = siteKey.userAccountId,
        ) ?: throw GlobalException.DataNotFound("Site not found")

        return siteKey
    }

    fun revokeActive(siteId: String) {
        val activeKey = siteKeyRepository.findBySiteIdAndStatusAndDeletedAtIsNull(
            siteId = siteId,
            status = SiteKeyStatus.ACTIVE,
        ) ?: return

        activeKey.status = SiteKeyStatus.REVOKED
        activeKey.revokedAt = Instant.now()

        siteKeyRepository.save(activeKey)
    }

    private fun generateRawKey(): String {
        val bytes = ByteArray(32)
        SecureRandom().nextBytes(bytes)

        return "nodus_live_" + Base64.getUrlEncoder()
            .withoutPadding()
            .encodeToString(bytes)
    }

    companion object {
        private const val KEY_PREFIX_LENGTH = 24
    }
}
